streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java

```diff
@@ -620,22 +620,35 @@ public final TopologyBuilder copartitionSources(Collection<String> sourceNodes)
      * Returns the copartition groups.
      * A copartition group is a group of source topics that are required to be copartitioned.
      *
+     * @param applicationId
      * @return groups of topic names
      */
-    public Collection<Set<String>> copartitionGroups() {
+    public Collection<Set<String>> copartitionGroups(final String applicationId) {
```

Comment:
For now i think point 3 is ok, though i would always call  `setApplicationId` in the `KafkaStreams` constructor as there is no harm in doing it and there are some public methods that need it to be set.
As for point 4 & the point about threading - i think we should leave this to another PR. I'd like to get this finished without too many more changes